Fix (enhancement) on 8 milion records index

johanpiet2 opened the pull request:
This two functions slows it down to 0.85ms. I changed it and it does 0.09ms

arElasticSearchInformationObjectPdo.class.php

  private function getCollectionRootId() //slows down to 0.224ms
  {
    //$sql = 'SELECT id FROM information_object force index (information_object_FI_17) 
    //  WHERE parent_id = ? AND lft < ? AND rgt > ?
    //';
	$sql = 'SELECT id FROM (SELECT * FROM information_object WHERE lft < ? AND rgt > ?  ) a WHERE a.parent_id = ?';

  return QubitPdo::fetchColumn($sql, array(QubitInformationObject::ROOT_ID, $this->lft, $this->rgt));

  public function getAncestors()
  {
    if (!isset($this->ancestors))
    {
      // Find ancestors
      $sql  = 'SELECT
                  node.id,
                  identifier,
                  repository_id,
                  slug';
      $sql .= ' FROM '.QubitInformationObject::TABLE_NAME.' node force index (information_object_FI_18)';
      $sql .= ' JOIN '.QubitSlug::TABLE_NAME.' slug
                  ON node.id = slug.object_id';
      $sql .= ' WHERE node.lft < ? AND node.rgt > ?';
      $sql .= ' ORDER BY lft';

      $this->ancestors = QubitPdo::fetchAll($sql, array($this->__get('lft'), $this->__get('rgt')));

information_object_FI_18 = index on lft and rgt

mvail committed:
Update EAD to PDF XSL task files, refs #10591

Updates to the XSL files that convert EAD to PDF, including:

- Increased logo size on Coverpage.
- Removal of "Rules for Archival Description" from Coverpage.
- Altered "generated by" on the Coverpage to read
  "Generated by <publisher> on <date>".
- Table of Contents now only has links to Series and Subseries.
  Removed links to files and items.
- Removed PDF book mark for files and items.
- Change the label "ID" to "Reference code".

mcantelon committed:
Context-sensitive D.O. browse link, refs #12797

Added a check, in the context menu of information object detail pages,
to see if a digital object exists before displaying a "Browse digital
objects" link.

mcantelon committed:
Add clipboard sending functionality, refs #12800

Add the ability to, if the feature's enabled and configured, send the
contents of the clipboard to an external script for processing.

A new clipboard settings page has been added and new settings related
to clipboard sending.

A parent class for settings pages has been added.

mcantelon committed:
Bump version to 2.6.0.

replaceafill committed:
Integrate MediaElement player, refs #12902

This commit replaces the flowplayer library with mediaelement for
playing video and audio digital objects.

It also updates the digital object code to create MPEG4 / H.264
reference derivatives instead of flash video (flv) derivatives.

replaceafill committed:
Derivative regeneration: media-type option, refs #12903

Added a command-line option, to the derivative regeneration task, that
filters based on the digital object media type ("audio" or "image" or
"text" or "video").

The confirmation message of the task has also been customized to this
effect.

rferris committed:
Convert UTC time to localtime in OAI requests, refs #13025

mcantelon committed:
Minor fixes to clipboard sending, refs #12800

Simplified clipboard sending message and hid clipboard send form fields
containing JSON.

sbreker committed:
Fix issue CLI DO load task, refs #12840

Fixed an issue in the DO load task where if the new --link-source option
is specified and a file in the spreadsheet cannot be found, it was
throwing an exception and stopping the load. I have updated this so
that it behaves the same as when the --link-source option is not
specified: it will log a message including the filename that was not
loaded and it will continue to process the import CSV without throwing
an exception.

Secondly if a file path specified in the import CSV was not found,
then the import task would create a stub information object without an
attached digital object (when the task is 'attaching' a digital object,
as opposed to 'linking'). This has been changed so that if a file from
the CSV is unable to be found, the task will not create a stub
information object record.

mcantelon committed:
Add clipboard send settings i18n, refs #12792

Added translation support to settings editing class and implemented
this capablility in clipboard settings child class.

replaceafill committed:
Video player fixes, refs #12902

* Preload video metadata

This helps the MediaElement player to get the aspect ratio before the
user clicks on the play button scaling it better.

* Fix audio/video DO loading from the full-width treeview

This loads the mediaelement assets in the index of the information
object so that the player works when the DOs are loaded through AJAX
calls from interacting with the full-width treeview.

It also converts the mediaelementplayer setup into a Drupal Behavior
(like the previous flowplayer was) so that it can be reattached when
the DOM is updated.

* Fix video thumbnail generation

This fixes the generation of thumbnails for videos shorter than 30
seconds by calculating the half of the video using ffprobe.

It also unifies the thumbnail generation logic with the one used by
the Flash uploader.

ross-spencer committed:
Init path as array instead of string, refs #13066

This commit corrects the initialization of path in importBulkTask to an array
to counter breaking changes introduced in PHP 7.1 that prevented its previous
initialization to be silently converted to an array.

replaceafill committed:
Video player fixes after testing, refs #12902

* Load mediaelement JS assets last

When the user is not authenticated the mediaelement assets are loaded
after the jQuery library breaking the player functionality.

This sets the mediaelement JS assets loading in the 'last' position.

* Set fixed size for mediaelement player

This removes the responsive stretching mode from the mediaelement
player setup and sets a 320x240 size used previously by the
flowplayer.

mcantelon committed:
Add clipboard send settings i18n, refs #12792

Fixed issues with translation support in settings editing class.

djjuhasz committed:
Set sfRoute->resource explicitly, refs #13069

The term/treeViewComponent (and possibly other components)
expect sfRoute->resource to provide the page resource context.

djjuhasz committed:
Add missing $resource parameter, refs #13044

replaceafill committed:
Fix copyright pop-up HTML, refs #13070

This prevents the content of the popup to be escaped by using the
Parsedown text method in unsafe mode.

djjuhasz committed:
Fix CSV import argument order, refs #13024

mcantelon committed:
Fix Firefox JS incompatibility, refs #12967

Fixed problem with navigation of global search form autocomplete
results, via the up and down arrows, in Firefox.

djjuhasz committed:
Reduce verbosity of ES update job, refs #13085

Reduce status updates to reporting after every 100 description updates to
avoid overflowing `jobs.output` column width.

djjuhasz committed:
Remove unnecessary data from ES, refs #13096

* Store only relevant repository data in the information object ES document.
* Store only relevant actor data for "name access point" relationship in
  information object ES document.
* Allow nesting _foreign_type mappings within a _partial_foreign_types mapping.

djjuhasz committed:
Truncate PDF text to <64KB, refs #13095

* Truncate digital object PDF text to <64KB to fit in MySQL TEXT field
* Truncate uploaded finding aid PDF text to <64KB too

mcantelon committed:
Add DACS element/header visibility, refs #13049

Added settings for visibility of DACS template headers and for the
physical access field element. Added notification when visibility
configuration is saved.

mcantelon committed:
Handle checkboxes in settings class, refs #13028

Added logic, to settings editing parent class, to handle checkbox
fields.

Make small change so child classes of settings editing parents class
don't have to define $I18N static property.

mcantelon committed:
Refactored DIP settings page, refs #13028

Refactored DIP settings page to use settings editing parent class.

sbreker committed:
New digitalobject load task option, refs #13103

Add new option to the CLI digitalobject:load CLI task: --attach-only

When set, this option will force all digital objects in the import CSV
file to be attached via their own information object record.

mcantelon committed:
Allow CSV import of I.O. translations, refs #8572

Made changes to enable import of translations.

mcantelon committed:
Refactored D.O. deriv. settings page, refs #13028

Refactored digital object derivatives settings page to use settings
editing parent class.

mcantelon committed:
Refactored security settings page, refs #13028

Refactored security settings page to use settings editing
parent class.

sevein committed:
Fix task plugin loading issue, refs #13109

This commit ensures that the action of disabling plugin loading in
tools:upgrade-sql is performed in a way that does not affect other
tasks.

tools:upgrade-sql disables application plugin loading by setting the
$loadPlugins global to false. However, that change was affecting all
tasks because it's performed in the configure method which is always
executed during the CLI bootstrap process, for every task, regardless
the task that the user is running.

A side effect was that jobs:worker was not able to access application
plugins like qtSwordPlugin unless the autoloading config had already
been cached by hitting the website first.

mcantelon committed:
Add warning to security settings, refs #13105

Added a warning to the security settings page to let users know that
incorrect security settings can result in the web UI becoming
inaccessible.

djjuhasz committed:
Gracefully handle missing user menus, refs #13113

- Fall back to old i18n labels if login, logout, or myProfile menus have
been deleted or renamed.
- Simplify display logic in template

djjuhasz committed:
Revert "Gracefully handle missing user menus..."

This reverts commit 856f9dd9c0402d8a3390d467c09d33d30bfd7a62.

mcantelon committed:
Fix full treeview paging issue, refs #13124

Fixed issue, for logged out users, with full treeview paging buttons not
appearing.

mcantelon committed:
Add warning when user toggles SWORD, refs #13117

Added a warning, on the plugin admin page, so that when an admin
enables or disables the SWORD plugin they are told to also restart
the AtoM worker.

mcantelon committed:
403 response for permission error, refs #12782

Changed the permission error page so it responds using a 403 HTTP code
rather than a 200 code.

mcantelon committed:
Fix finding aid custom L.O.D. issue, refs #12707

Fixed issue with finding aid generation where the name of custom
levels of description wouldn't be displayed.

djjuhasz committed:
Gracefully handle missing user menus, refs #13113

- Fall back to old i18n labels if login, logout, or myProfile menus have
  been deleted or renamed.
- Simplify display logic in template

jraddaoui committed:
Escape entities in hidden inputs, refs #13125

When Markdown support is enabled, the escaping strategy gets disabled.
This may cause issues in some hidden fields where the user input is
added without validation. The keys added to those inputs are already
validated in the controllers, but the values need to be escaped.

sbreker committed:
New digitalobject load task option, refs #13126

Add a new option to the digital object load task to disable nested
set value updates on import.

mcantelon committed:
Fix full treeview button visibility, refs #13124

Buttons, for full treeview button navigation, were showing when they
shouldn't have been.

jraddaoui committed:
Update Docker Compose environment, refs #13119

- Move Dockerfile to project root folder.
- Use project folder instead of tarball on build.
- Use `php:7.2-fpm-alpine` as base image.
- Remove `mcrypt` PHP extension.
- Install `npm` instead of `nodejs`.
- Unify RUN layers in Dockerfile.
- Improve ports configuration.
- Use default network.
- Use environment file.
- Use default driver for named volumes.
- Remove command from atom service.
- Upgrade Percona and Elasticsearch images.
- Avoid ES errors on low space disks.
- Fix and enhance MySQL configuration.
- Improve atom_worker stability.
- Rename Nginx configuration file.
- Remove push-image script.
- Cosmetic changes.

Squashed commit of the following:

commit 2df9996aea35b6407ee75477255bd9e0b7214f07
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Wed Jul 31 19:13:52 2019 +0200

    Unify RUN layers in Dockerfile

commit 992ebd8ed601f3a566a68db7713bbed8d57b5079
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Wed Jul 31 18:09:10 2019 +0200

    Remove push-image script

commit d0e5441a260a1ea9574a15f75417326193f01cbe
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 21:00:47 2019 +0200

    Cosmetic change: remove unneeded quotes

commit c0a347d968b206362e6ed60adbf39e540551feab
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 20:59:08 2019 +0200

    Rename Nginx configuration file

commit 64e1d467c7c48a7201eb14b7180e091267d40de5
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Wed Jul 31 15:08:48 2019 +0200

    Improve atom_worker stability

    Start the container after Gearmand and MySQL and launch the worker
    command after sleeping 10 seconds to reduce the possibility of error
    connecting to those services. Restart on failure (max 5) which also
    sleeps those 10 seconds in between restarts.

commit eea1c6a9ae764653254fcbdf3c3c5dee126a232a
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 20:57:17 2019 +0200

    Fix MySQL configuration

    - The configuration folder has been moved in `percona:5.7` to a
      different location inside the container: `/etc/my.cnf.d`.
    - Unify encoding and sql_mode configuration files.
    - Concatenate `init_connect` statements.
    - Include `STRICT_TRANS_TABLES` in `sql_mode` (as in documentation).
    - Add `optimizer_switch` (as in documentation).

commit ecdabb4d23a06291f928e1edf840f26ea283dc53
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 14:50:45 2019 +0200

    Cosmetic changes

commit 9700fb6460ba1a63fe9ce6f7e5b4444fe3a5bb5e
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 14:49:55 2019 +0200

    Cosmetic changes

commit de063d612ddb4b7a0af58404313e1d7d38f6bcea
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 14:48:09 2019 +0200

    Avoid ES errors on low space disks

    This is calculated with a percentage, which may cause problems in
    development environments with over 85-90% of the disk space used.

commit 9b7ddc0f80080b80d5d88ed5cb79fa77cc64e0e1
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 14:42:50 2019 +0200

    Upgrade Elasticsearch image

commit 3bd67d7027bc958f7e15d875d7c4878c69510e15
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 14:26:34 2019 +0200

    Upgrade Percona image

commit 19755b6cbcf7a6d4b0a126cffee4d16beec08502
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 14:25:24 2019 +0200

    Use environment file

commit 1c65174cad7d915cb84f2d1b4c5ca8ddff9098d1
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 13:53:45 2019 +0200

    Fix volume path

    To keep consistency with the ones changed in the atom and atom_worker
    services. This is just a cosmetic change.

commit 04246b2b9f6120d9c5d621fd634a3b19940e2a93
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 13:44:36 2019 +0200

    Remove command from atom service

    The command used in the Dockerfile is already `fpm`.

commit 784caae1c8c4ff8413255d2acff7d97bfebf843a
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 13:43:09 2019 +0200

    Use default driver for named volumes

     Defaults to whatever driver the Docker Engine has been configured
     to use, which in most cases is local.

commit 5beb826c98aaf2728a2ae186487ee82f6892b97d
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 13:41:05 2019 +0200

    Use default network

    Compose sets up a single network for the environment by default. Each
    container for a service joins the default network and is both reachable
    by other containers on that network, and discoverable by them. Since
    there is no limitation on the services interaction and we were using
    "default" networks, this simplifies the setup.

commit 1d839f3a6b5412100d2b6b4b7345a84defe85fae
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 13:29:51 2019 +0200

    Improve ports configuration

    - The Elasticsearch image already exposes 9300.
    - Use ports instead of expose and map host/container ports.
      - Limit internal services to the host.
      - Allow access to Nginx port from outside the host.

commit 8da91e902d909709f9657e06fc6949150b7ccfdd
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 13:11:07 2019 +0200

    Use php:7.2-fpm-alpine as base image

    - Remove `mcrypt` PHP extension.
    - Install `npm` instead of `nodejs` as they come separated now.

commit c34b84678f62457c5caf670dfaddc3d2c961e3c1
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Jul 30 13:07:12 2019 +0200

    Move Dockerfile to project root folder

    This allows to copy the project folder in its current status instead
    of declaring a branch argument and fetching a tarball from Github.

sbreker committed:
Fix options digitalobject load task, refs #13103

This fixes an issue with the digitalobject:load task where some options
were not being respected.

mcantelon committed:
Round trip mode for CSV import, refs #12281

Added round trip mode to CSV information object import. Round trip mode
treats the values in the legacyID column as internal AtoM IDs and
doesn't create keymap entries.

jraddaoui committed:
Use config to generate jobs prefix, refs #13108

Add `workers_key` setting to `config/app.yml` and use it alongside the
`sf_root_dir` to generate the job prefix instead of the site title and
base URL.

The site title and base URL were used to generate the job prefix, which
leads to different prefixes when those values are changed in the
database or when using a different language. If the prefix changes
between the AtoM worker start and when a job is triggered, no worker
will be found to handle that job.

This prefix was introduced to avoid collisions between workers from
different instances connected to the same Gearman server. However, if
those instances are running on the same machine, the `sf_root_dir` will
be different on each of them. Therefore, using the title and URL only
prevents the collision when multiple instances are running on different
machines, on the same path and they're connected to the same Gearman
server.

This is not an usual scenario and requires to modify configuration files
manually, at least `config/gearman.yml` in some of those instances to
point to a different server. Therefore, setting a different value for
the `workers_key` on each instance seems like a good solution to avoid
the collisions.

In the majority of the deploys, even in those where multiple AtoM
instances are running on the same machine, this new setting can be left
empty.

jraddaoui committed:
Get raw parameters in filter tags, refs #13131

When the escaping strategy is enabled and Markdown support is disabled,
`getParams` is converted to a `sfOutputEscaperArrayDecorator` instance.
This needs to be converted back using `$sf_data->getRaw()` to be able
to merge it with another array in the template.

jraddaoui committed:
Fix arDominionPlugin theme header, refs #13137

Make sure the content of the site header is always wrapped within it.
Use a pseudo element and clearance to avoid disrupting themes extending
this one. This also makes the header play nicer with logos higher than
50px and long titles, fixing #3895 and #6293 too.

mcantelon committed:
Optional full treeview collapse, refs #13135

Add, optionally, a collapsible area around the full width treeview on
description pages.

The text shown on the collapse "button" can be set by an admin on the
user interface label settings page.

jraddaoui committed:
Fix status code in API responses, refs #13143

jraddaoui committed:
Fix creation of DOs without path, refs #13144

This issue came out now probably because of the upgrade to MySQL 5.7
and/or the re-addition of `STRICT_TRANS_TABLES` to the `sql_mode`.

jraddaoui committed:
Fix DO metadata component, refs #13145

When Markdown is disabled the escaping strategy converts
`$resource->object` to an instance of `sfOutputEscaperObjectDecorator`,
which fails to match `QubitInformationObject` when it's compared in the
template. Check the related resource type in the component action
before the transformation happens.

mcantelon committed:
Render repo description as Parsedown, refs #13106

Render the repo description HTML field as Parsedown rather than just
HTML.

jraddaoui committed:
DIP upload fixes, refs #13067 and #13068

Squashed commit of the following:

commit 4764110f9426d4fc288b2e64bd185a7fb3d0c053
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Aug 12 15:26:15 2019 +0200

    Fix all DMD section finding

    The DMDID attribute can contain one or more DMD section ids
    (e.g.: DMDID="dmdSec_2 dmdSec_3"). When multiple DMD sections
    are associated with the same file/dir we'll try to return the
    latest one created.

commit 253d196f82a704ae76366e746d22109979eff98e
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Aug 9 22:24:40 2019 +0200

    Fix main DMD section finding

    Use always the physical (default) structMap.

commit c8dda44b29bbb56a9c8c00c00ef148af177c3a40
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Aug 9 20:47:40 2019 +0200

    Allow different namespaces in DC metadata

    Allow no namespace and `dc` or `dcterms` namespaces.

commit 06e7f08d64e51607c9639fd4e449976bbe8fbb80
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Aug 9 20:45:04 2019 +0200

    Get namespaces from METS file when possible

    Use `getDocNamespaces` to get the namespaces declared in the
    METS file dynamically from the document. To keep backwards
    compatibility, add the old URI for main namespaces if they're
    not found in the document.

    Use a key/name mapping to add those namespaces in XPath queries
    without modifying the existing ones.

fiver-watson committed:
Add new SECURITY.md, refs #13139

fiver-watson committed:
Minor fixes to .md files, refs #13139

mcantelon committed:
Refactor privacy notif. setting page, refs #13028

Refactored privacy notification settings page to use settings editing
parent class.

fiver-watson committed:
Update README, refs #13155

fiver-watson committed:
fixup of README

mcantelon committed:
Update keyboard handling events in JS, refs #12725

Updates keyboard handling events to remove use of the deprecated
"keypress" event and "keyCode" event properties replacing these,
respectively, with "keydown" events and "which" event properties.
jQuery populates the "which" event property even in browsers that
don't support "which" natively.

jraddaoui committed:
Escape entities, refs #13125

- The start and end dates from the adv. search were used to form the
  filter tag label directly from user input.
- The inline-search query was not escaped, this will normally raise an
  ES error before the template is rendered, but it may be an issue if
  the "Escape special chars from searches" setting avoids that error.

mcantelon committed:
Fix CSV import of I.O. translations, refs #8572

Skip importing related data on rows than contain a translation.

mcantelon committed:
Add authority record CSV export task, refs #13148

Add task to export authority record data as CSV file(s). If exporting
into a directory then aliases and relations can also be exported.

sbreker committed:
Accessions import enhancements, refs #12799

This commit adds the ability to import a new CSV column donorNotes which
will import to the accession donor notes field which already exists.

Secondly, the import task has been changed such that if a new
accession material type term or acquisition type term is encountered in
the import CSV, the task will now create the new term and log a message
instead of throwing an exception and stopping the import.

I have also updated the log message that is created when a duplicate
accession identifier is found to include the accession identifier.

mcantelon committed:
Update translations in treeview AJAX, refs #8572

When navigating from description to description, using the full
treeview, the select element, for choosing an available translation,
wasn't getting updated. Fixed.

mcantelon committed:
CSV translation import: update index, refs #8572

Added search index updating during CSV import of information object
translations.

mcantelon committed:
Fix base URL in OAI identify XML, refs #13163

When AtoM's base URL was set to an "https" URL the identify response
was changing this to an "http" URL. Fixed and simplified.

fiver-watson committed:
Update storage perms, refs #13169

mcantelon committed:
Truncate stored job output to 64K, refs #13174

Added logic to limit stored job output to 64K.

mcantelon committed:
Update references to renamed JS file, refs #13177

djjuhasz committed:
Improve alphabetic sort, refs #5173

- Add custom 'alphasort' normalizer
- Apply 'alphasort' normalizer to special 'alphasort' ES field
- Add alphasort fields for accession, actor, information object,
  repository, term
- Use alphasort fields for ES sorting where appropriate

mcantelon committed:
Remove identifier Markdown parsing, refs #13173

Refactored display of title, etc., in a number of description standard
plugins and got rid of Markdown parsing of identifiers.

* Created a component to display level of description, reference code,
  title, and publication status
* In the component the reference code doesn't get parsed as Markdown,
  but the title does
* Refactored description standard plugins to use component
* Removed Markdown parsing for misc. identifiers in index/edit page
  templates

mcantelon committed:
Add creator inheritance to DC XML, refs #13130

Changed the DC XML component so creators will be inherited from
ancestors.

mcantelon committed:
Refactor site information settings, refs #13028

Refactored the site information settings page to use the settings
editing parent class.

sbreker committed:
Related term ES index speed, refs #13175

This commit contains an enhancement to speed up ES indexing of
information objects when there is a large number of terms in the terms
table.

- Subject, place and genre term ids and parent_ids are preloaded into an
array so a term's parent can be found directly using the term id as the
array index.

For each directly related term, parents are found by searching the term
array recursively instead of using the nested set ancestors query using
the LFT and RGT columns.

This change results in a large performance increase when indexing
information objects when testing with ~350K place and subject terms.

- Changed related terms query to use a subquery for performance. The old
query would take 12 seconds in some cases. The subquery technique is
faster in all cases.

- Load the terms lookup table straight from PDO using PDO::FETCH_KEY_PAIR
option. This removes the need to build the lookup table by hand.

Testing shows a performance increase when using PDO::FETCH_KEY_PAIR over
building the lookup table by hand.

djjuhasz committed:
Add physical object CSV import, refs #13183

- Add csv:physicalobject-import symfony task
- Support --culture, --error-log, --header, --index, --rows-until-update,
  --skip-rows, and --source-name options
- Use composer to manage dependencies
- Use http://csv.thephpleague.com library for CSV file parsing
- Add PHPUnit 8 unit tests for new code
- Use vfsStream library for mocking file system in tests

mcantelon committed:
Fix D.O. metadata header display, refs #13190

Fixed issued where, if Markdown is disabled, digital object metadata
section header doesn't display.

jraddaoui committed:
Do not disable escaping strategy, refs #13192

To avoid issues rendering Markdown, unescape the content only in the
Markdown parsing process, which is normally used in safe mode. This
reduces performance (escaping and unescaping the content) but increases
security (having all user inputs escaped by default).

mcantelon committed:
AJAX paging for actor event editing, refs #13157

Use AJAX to address scalability issues when an actor has a large amount
of events and an AtoM user attempts to edit the actor. Added the
ability to handle dynamically added content to js/multiDelete.js.

Used the "prettier" tool to format some related JS files.

mcantelon committed:
Add treeview AJAX translation links, refs #8572

When navigating from description to description, using the full
treeview, if translation links weren't present in the current page
then the translation links weren't getting updated. Fixed.

sbreker committed:
Accession extent multiline edit, refs #13198

Change the accession edit page to accept multiline input for the
"Received extent units" field.

mcantelon committed:
Fixes to header display component, refs #13173

Fixed issues in the header display component that was added to fix
Markdown parsing issues.

Issues fixed:

* A function signature in the component was incorrect
* The component's template filename conflicted with with site header
  template filename causing an inappropriate template override

mcantelon committed:
Fix DC XML creator inheritance, refs #13130

Changed the DC XML component so creators can be inherited from non-top
level information objects.

sbreker committed:
Allow 'add' menu access, Refs #13202

In addition to restricting AtoM's 'add' menu to specific hard coded
groups, check ACL groups to see if the 'add' menu should appear for
users in custom ACL groups. Previously, custom groups were not being
checked to determine if AtoM's 'add' menu should be displayed for
these users.

Removed term, Isdf and Isdiah security.yml files to allow access to
these modules to authenticated users in custom groups.

Co-Authored-By: Mike Cantelon <mcantelon@gmail.com>

jraddaoui committed:
Add actor:move-description-relations, refs #13189

Add a new task to move the description relations from one actor to
another. Including all events and subject access point relations.
Include a skip-index option to avoid updating the ES index.

mcantelon committed:
Add warning to CSV I.O. I18n import, refs #8572

Prevent attempts to add related data during CSV import of information
object translations.

Add a warning to CSV import of information object translations when
columns that aren't supported are populated.

djjuhasz committed:
Use culture fallback for location, refs #13201

Display source culture in physical object browse list when an i18n value
for the current culture is not available.

mcantelon committed:
Fix to finding aid custom LOD issue, refs #12707

Fixed issue with finding aid generation: name of custom levels of
description wouldn't be displayed.

djjuhasz committed:
Move composer autoload to importer, refs #13183

Requiring `vendor/composer/autoload.php` in ProjectConfiguration was causing an
application-wide critical error if the autoloader wasn't installed via
composer.

jraddaoui committed:
Fix CSV notes import, refs #13209

Check `SOURCE_CULTURE` constant instead of `sourceCulture` property
as the latter is accessed through magic methods and `property_exits`
can't find it.

fiver-watson commented:
Hello @johanpiet2 , thank you for submitting this! Unfortunately, you have opened this PR against our stable/2.5.x branch, and it looks like the PR currently includes 85 commits because of this. Would you mind closing this, and opening a new pull request against our qa/2.6.x branch, after you have rebased your work against that branch, so that your PR includes only the commits related to your ES index improvements? Thank you in advance!

mcantelon committed:
Fix term view page search issue, refs #13182

Fixed incorrect function call to fix searches done from the term view
page's sidebar.

jraddaoui committed:
Improve IO hierarchy deletion, refs #13211

Avoid nested set update until the deletion of the top-level description
in the tree. The queries used to update the nested may be time expensive
as they update all the descriptions above the deleted description,
including those outside the deleted tree and those that are inside and
will be deleted after. When the `deleteFromNestedSet` function from
`BaseInformationObject` is called, the delta used to update the LFT and
RGT values is calculated from the resource's RGT -LFT difference.
Considering that this operation is normally run inside a transaction
(otherwise it should), updating the nested set only once at the end
should be enough.

djjuhasz committed:
CSV update of physical objects, refs #13185

* Add --update option to csv:physicalobject-import to update existing physical
  object data via CSV import
* Add --skip-unmatched option to prevent the addition of new physical objects
  during update
* Add --multi-match option to control which physical objects to update if more
  than one matches the import name (options: skip, first, all)
* Add --partial-match option to match and update existing physical objects when
  the import name matches the beginning of an existing PO name
* Add --empty-overwrite option to allow empty CSV values to overwrite existing
  DB data on update
* Update unit tests

djjuhasz committed:
Add composer deps when building tarball

- Install phing with composer (instead of old PEAR channel)
- Add composer production dependencies to tarball build
- Add additional phing build messages for <exec> steps to aid debugging
- Bump composer managed library versions

mcantelon committed:
Fix clipboard send escaping issue, refs #12800

Fixed escaping-related issue with the clipboard send feature.

mcantelon committed:
Update password hashing to Argon2, refs #12136

Updates the hashing logic from using SHA1 to using Argon2. To avoid
users having to re-enter their passwords we use Argon2 to hash the
original SHA1 hash of the user's password (plus salt). As Argon2
automatically generates the salt no additional salt-related logic is
needed.

The hashing algorithm and the options used with it can be changed by
editing the config/app.yml file and changing the
password_hash_algorithm and/or password_hash_algorithm_options options.

The hash algorithm and options are stored as part of the hash string
data so changing hash algorithm and options won't stop previously
stored password hashes from working.

mcantelon committed:
Change actor edit pager button text, refs #13157

Changed the text for the actor edit event pager's next button from
"More" to "Show more" for greater visibility.

mcantelon committed:
Full treeview fix and paging setting, refs #13186

Fixed issue with full treeview not scrolling to the active treeview
node. Made the number of nodes shown, per treeview page, configurable
via a setting.

fiver-watson commented:
As previously noted, this PR was created incorrectly, and does not actually include the commit with the ES indexing enhancement. Currently it is grabbing every commit added to qa/2.6.x and trying to merge it into our stable/2.5.x branch. Because of this, I'm now closing this PR. @johanpiet2 I hope you resubmit your change as a single commit against qa/2.6.x so we can review it!
